### PR TITLE
[SPARK-24351][SS]offsetLog/commitLog purge thresholdBatchId should be computed with current committed epoch but not currentBatchId in CP mode

### PR DESCRIPTION
**Pull request by ivoson**
## What changes were proposed in this pull request?
Compute the thresholdBatchId to purge metadata based on current committed epoch instead of currentBatchId in CP mode to avoid cleaning all the committed metadata in some case as described in the jira [SPARK-24351](https://issues.apache.org/jira/browse/SPARK-24351).

## How was this patch tested?
Add new unit test.

### CODE REVIEW

**Comment by ivoson** [comment]
hi, @xuanyuanking @jose-torres can you take a look at this pr please?

**Comment by jose-torres** [comment]
Nice catch!

Consider adding a unit test for this. I'm not sure how easy it'd be off the top of my head to test what the offset log is retaining, and I think it's a simple enough change to do without if needed.

**Comment by ivoson** [comment]
@jose-torres thanks for reply. I will try to add a unit test for this.

**Comment by ivoson** [comment]
hi @jose-torres,  the last commit add a unit test for this.

**Review by xuanyuanking — COMMENTED** [review]
+1 for me, cc @jose-torres @zsxwing. 
We found this bug during long running CP shuffle practice in Baidu, while shuffle mode take more time and more easier to trigger this bug.

**Comment by jose-torres** [comment]
LGTM other than the one comment. Sorry it took me so long to get to this.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 1]
```diff
@@ -34,7 +34,8 @@ class ContinuousSuiteBase extends StreamTest {
     new SparkContext(
       "local[10]",
       "continuous-stream-test-sql-context",
-      sparkConf.set("spark.sql.testkey", "true")))
+      sparkConf.set("spark.sql.testkey", "true")
+        .set("spark.sql.streaming.minBatchesToRetain", "2")))
```
[the remark this answers is not in the document]
Comment (ivoson) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: @jose-torres Thanks for pointing this out. I will make a separate suite for clarity and isolation.

**Comment by ivoson** [comment]
cc @jose-torres  Thanks for the review. The last commit addresses the comments you left.

**Comment by jose-torres** [comment]
LGTM

**Comment by zsxwing** [comment]
ok to test

**Comment by zsxwing** [comment]
LGTM pending tests

**Comment by zsxwing** [comment]
Thanks! Merging to master

**Comment by ivoson** [comment]
@jose-torres @xuanyuanking  @zsxwing Thanks for reviewing this.

**Pull request closed without merging**